src/main/java/uk/nhs/hee/tis/trainee/credentials/service/GatewayService.java

```diff
@@ -149,4 +151,76 @@ private Optional<URI> buildCredentialUri(ResponseEntity<ParResponse> response) {
   record ParResponse(@JsonProperty("request_uri") String requestUri) {
 
   }
+
+  /**
+   * Call a token code endpoint and extract the claims from the resulting token.
+   *
+   * @param endpoint     The token endpoint.
+   * @param redirectUri  The redirect endpoint sent with the initial request.
+   * @param code         The code resulting from the initial request.
+   * @param codeVerifier The code PKCE code verifier matching the challenge sent with the initial
+   *                     request.
+   * @return The extracted claims.
+   */
+  public Claims getTokenClaims(URI endpoint, URI redirectUri, String code, String codeVerifier) {
+    String state = UUID.randomUUID().toString();
+    HttpEntity<MultiValueMap<String, String>> request = buildTokenRequest(redirectUri, code,
+        codeVerifier, state);
+
+    log.info("Sending token request.");
+    ResponseEntity<TokenResponse> tokenResponse = restTemplate.postForEntity(endpoint, request,
+        TokenResponse.class);
+
+    if (tokenResponse.getStatusCode().isError()) {
+      log.error("Token request failed with code {}.", tokenResponse.getStatusCode());
+      return new DefaultClaims();
+    }
+    log.info("Received token response.");
+
+    TokenResponse body = tokenResponse.getBody();
+
+    if (body == null) {
+      log.error("Token response was empty.");
+      return new DefaultClaims();
+    }
+
+    // TODO: verify token against public key
+    String signedToken = body.idToken();
+    return jwtService.getClaims(signedToken);
+  }
+
+  /**
+   * Build a request to get a token for the given code.
+   *
+   * @param redirectUri  The redirect endpoint sent with the initial request.
+   * @param code         The code resulting from the initial request.
+   * @param codeVerifier The code PKCE code verifier matching the challenge sent with the initial
+   *                     request.
+   * @param state        The state to include in the request.
+   * @return The built token request.
+   */
+  private HttpEntity<MultiValueMap<String, String>> buildTokenRequest(URI redirectUri, String code,
+      String codeVerifier, String state) {
+    log.info("Building Token request.");
+
+    MultiValueMap<String, String> bodyPair = new LinkedMultiValueMap<>();
+    bodyPair.add("client_id", properties.clientId());
+    bodyPair.add("client_secret", properties.clientSecret());
+    bodyPair.add("redirect_uri", redirectUri.toString());
+    bodyPair.add("grant_type", "authorization_code");
+    bodyPair.add("code", code);
+    bodyPair.add("code_verifier", codeVerifier);
```

Comment:
This is not needed for retrieving the token for an issued credential; I assume it will be ignored.

Reply:
It **should** be needed if we send a code challenge with the initial request, I'm assuming that's not currently the case?

Reply:
Not as far as I know. I can't see any way to do this in the swagger docs.

Reply:
I had a play around with Postman, it doesn't look like PKCE is implemented for the issue endpoints, That does kind of make sense, since we don't care so much about them.
If the param is passed with the request it's just ignored though, but we could make this conditional too if we needed to.